### PR TITLE
[video_player] Migrate video player to null safety

### PR DESCRIPTION
**Pull request by blasten**
Migrates video player to null safety.

I made some API changes like the use of `Duration.zero`/`Size.zero` instead of `null` for defaults in `VideoPlayerValue` and `Caption`.

### CODE REVIEW

File: packages/video_player/video_player/lib/video_player.dart [hunk 1]
```diff
@@ -41,17 +42,20 @@ class VideoPlayerValue {
...
-      : this(duration: null, errorDescription: errorDescription);
+      : this(
+            duration: Duration.zero,
+            isInitialized: false,
+            errorDescription: errorDescription);
 
   /// The total duration of the video.
   ///
-  /// Is null when [initialized] is false.
+  /// The duration is zero if the video hasn't been initialized.
```
Comment (cyanglaz) [inline packages/video_player/video_player/lib/video_player.dart:58]: nits: 
```suggestion
  /// The duration is [Duration.zero] if the video hasn't been initialized.
```

File: packages/video_player/video_player/lib/video_player.dart [hunk 2]
```diff
@@ -84,15 +88,15 @@ class VideoPlayerValue {
...
+  final String? errorDescription;
 
   /// The [size] of the currently loaded video.
   ///
   /// Is null when [initialized] is false.
   final Size size;
 
   /// Indicates whether or not the video has been loaded and is ready to play.
-  bool get initialized => duration != null;
+  final bool isInitialized;
```
Comment (cyanglaz) [inline packages/video_player/video_player/lib/video_player.dart:99]: Why did we need to make this name change? Feels like we can make it as non-breaking as possible?
Although I do agree that `isInitialized` is a better name. So your call.

File: packages/video_player/video_player/lib/video_player.dart [hunk 3]
```diff
@@ -340,19 +356,18 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
   }
 
   @override
+  // ignore: must_call_super
```
Comment (cyanglaz) [inline packages/video_player/video_player/lib/video_player.dart:359]: why remove super?

File: packages/video_player/video_player/lib/video_player.dart [hunk 4]
```diff
@@ -340,19 +356,18 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
...
         _isDisposed = true;
         _timer?.cancel();
         await _eventSubscription?.cancel();
         await _videoPlayerPlatform.dispose(_textureId);
       }
-      _lifeCycleObserver.dispose();
     }
-    _isDisposed = true;
-    super.dispose();
+    _lifeCycleObserver.dispose();
```
Comment (cyanglaz) [inline packages/video_player/video_player/lib/video_player.dart:370]: Looks like the logic here is changed, is it related to null-safety migration?

File: packages/video_player/video_player/lib/video_player.dart [hunk 5]
```diff
@@ -400,11 +415,11 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
           if (_isDisposed) {
             return;
           }
-          final Duration newPosition = await position;
+          final Duration? newPosition = await position;
           if (_isDisposed) {
             return;
           }
-          _updatePosition(newPosition);
+          _updatePosition(newPosition!);
```
Comment (cyanglaz) [inline packages/video_player/video_player/lib/video_player.dart:422]: can `newPosition` here be null?

File: packages/video_player/video_player/lib/video_player.dart [hunk 6]
```diff
@@ -545,7 +560,7 @@ class _VideoAppLifeCycleObserver extends Object with WidgetsBindingObserver {
   final VideoPlayerController _controller;
 
   void initialize() {
-    WidgetsBinding.instance.addObserver(this);
+    WidgetsBinding.instance!.addObserver(this);
```
Comment (cyanglaz) [inline packages/video_player/video_player/lib/video_player.dart:563]: Can `WidgetsBinding.instance` be null?

File: packages/video_player/video_player_web/lib/video_player_web.dart [hunk 7]
```diff
@@ -66,21 +66,22 @@ class VideoPlayerPlugin extends VideoPlayerPlatform {
...
-        if (dataSource.package != null && dataSource.package.isNotEmpty) {
+        String assetUrl = dataSource.asset!;
+        if (dataSource.package != null && dataSource.package!.isNotEmpty) {
           assetUrl = 'packages/${dataSource.package}/$assetUrl';
         }
         // 'webOnlyAssetManager' is only in the web version of dart:ui
         // ignore: undefined_prefixed_name
-        assetUrl = ui.webOnlyAssetManager.getAssetUrl(assetUrl);
+        ui.webOnlyAssetManager.getAssetUrl(assetUrl);
+        assetUrl = '';
```
Comment (cyanglaz) [inline packages/video_player/video_player_web/lib/video_player_web.dart:84]: Not sure why we need this line, could you elaborate?

File: packages/video_player/video_player/lib/video_player.dart (hunk 1 above)
Comment (blasten) [reply packages/video_player/video_player/lib/video_player.dart]: Done.

File: packages/video_player/video_player/lib/video_player.dart (hunk 2 above)
Comment (blasten) [reply packages/video_player/video_player/lib/video_player.dart:99]: Ack.

File: packages/video_player/video_player/lib/video_player.dart (hunk 3 above)
Comment (blasten) [reply packages/video_player/video_player/lib/video_player.dart]: This was bad merge. Fixed.

File: packages/video_player/video_player/lib/video_player.dart (hunk 4 above)
Comment (blasten) [reply packages/video_player/video_player/lib/video_player.dart]: Fixed.

File: packages/video_player/video_player/lib/video_player.dart (hunk 5 above)
Comment (blasten) [reply packages/video_player/video_player/lib/video_player.dart]: No, but I changed the code, so it's more readable. The Analyzer can figure out that `newPosition!` can safely be ``newPosition` if you have a null check earlier.

File: packages/video_player/video_player/lib/video_player.dart (hunk 6 above)
Comment (blasten) [reply packages/video_player/video_player/lib/video_player.dart:564]: no. `runApp()` calls `WidgetsFlutterBinding.ensureInitialized()`.

File: packages/video_player/video_player_web/lib/video_player_web.dart (hunk 7 above)
Comment (blasten) [reply packages/video_player/video_player_web/lib/video_player_web.dart]: my mistake. Fixed.

**Review by cyanglaz — APPROVED** [review]
lgtm

**Pull request merged**